### PR TITLE
[Car Racing] Submit Jihyun Lim's mission

### PR DESCRIPTION
**Pull request by Jiihyun**
## **Let's Grow Together Through Code Reviews!**

Feel free to leave any comments or feedback on this PR! I’ll also make sure to review your PRs as well. I believe that by reviewing each other’s code, we can learn from different perspectives and grow together.
Don’t hesitate to share your thoughts! 😊

## What I Considered
- **Introducing the `NumberGenerator` Interface and `FixedNumberGenerator` Class**
  - To test functionality dependent on random number generation, I implemented the `NumberGenerator` interface and a `FixedNumberGenerator` class. This allowed me to decouple the randomness from the core logic and simulate predictable outputs for testing.
- **Wrapping Primitive Values (e.g., `Position`, `Name`, `Round`)**
  - By wrapping primitive values such as `Position`, `Name`, and `Round` into dedicated objects, I was able to add meaningful constraints, enforce validation rules, and clearly define their roles within the domain.
- **Creating Immutable Objects**
    - By creating immutable objects for entities like `Cars` and `Car`, I was able to ensure that the state of each round remains unchanged, enhancing data integrity and simplifying debugging. This approach also allowed for safely storing race progress in `RaceHistory` without unintended side effects.

## What I'd Like to Discuss with Reviewers
- Whether I effectively **utilized Kotlin's APIs** (Note: This is my first Kotlin project😁)
- I tried to assign clear responsibilities to each object to achieve both object-oriented programming principles and readable code. 
I’m curious to hear your thoughts on **whether this approach is reflected** in my mission.
- I noticed Kotlin's type inference can sometimes make code harder to understand.
I'm curious about how others approach this—**do you prefer to explicitly declare local variable types or rely on type inference?**

## Package Structure Overview

```
racingcar
    ├── Application.kt                            # Main entry point
    ├── controller
    │   └── RacingController.kt                   # Manages the flow of the car racing
    ├── domain
    │   ├── Referee.kt                            # Determines the winner
    │   ├── car
    │   │   ├── Car.kt                            # Represents a single car
    │   │   ├── Name.kt                           # Validates and stores car name
    │   │   └── Position.kt                       # Tracks car position
    │   ├── numbergenerator
    │   │   ├── NumberGenerator.kt                # Interface for generating numbers
    │   │   └── RandomNumberGenerator.kt          # Generates random numbers
    │   └── race
    │       ├── Cars.kt                           # Manages a collection of cars
    │       ├── Race.kt                           # Handles race progression
    │       ├── RaceHistory.kt                    # Tracks race history
    │       ├── RaceResult.kt                     # Stores race results
    │       └── Round.kt                          # Tracks total rounds
    ├── util
    │   └── StringParser.kt                       # Parses inputs
    └── view
        ├── InputView.kt                          # Handles user input
        └── OutputView.kt                         # Displays output

```

### CODE REVIEW

File: src/main/kotlin/racingcar/util/StringParser.kt [hunk 1]
```diff
@@ -0,0 +1,18 @@
+package racingcar.util
+
+class StringParser private constructor() {
+
+    companion object {
+        private const val INVALID_NUMBER_ERROR = "[ERROR] Input value must be a number."
+
+        fun parseByDelimiter(input: String, delimiter: String): List<String> {
```
Comment (puleugo) [inline src/main/kotlin/racingcar/util/StringParser.kt:8]: I think we can simplify this by using `default parameter`.
Currently, when you call this method, you're injecting a value into a `private static field`. This means that someone reading the code for the first time will **need to scroll to the bottom of the class and then come back** to understand what's happening.
If you use a `default parameter` like below, it will make the code easier to read — unless there's a specific reason not to.

```suggestion
        fun parseByDelimiter(input: String, delimiter: String = ","): List<String> {
```

File: src/main/kotlin/racingcar/controller/RacingController.kt [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+        val round = initRound()
+        val result = playRace(round, cars)
+        determineWinner(result)
+    }
+
+    private fun initCars(): Cars {
+        val carNames = inputView.readCarNames()
+        val cars = carNames.map { Name(it) }
+            .map { Car(it) }
+            .toList()
```
Comment (puleugo) [inline src/main/kotlin/racingcar/controller/RacingController.kt:28]: I think we can simplify this by remove `.toList()` method.
```suggestion
```
The `.map()` method's result is already `List`.

File: src/main/kotlin/racingcar/controller/RacingController.kt [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+    private val outputView: OutputView
+) {
+    fun run() {
+        val cars = initCars()
+        val round = initRound()
+        val result = playRace(round, cars)
+        determineWinner(result)
+    }
+
+    private fun initCars(): Cars {
```
Comment (puleugo) [inline src/main/kotlin/racingcar/controller/RacingController.kt:24]: Could you consider renaming the `initCars()` method to something more descriptive, like `readCarsFromInput()`?
`initCars()` is a bit unclear.

File: src/main/kotlin/racingcar/domain/race/RaceHistory.kt [hunk 4]
```diff
@@ -0,0 +1,16 @@
...
+
+    fun recordRound(cars: Cars) {
+        history.add(cars)
+    }
+
+    fun getRaceResult(): RaceResult {
+        return RaceResult(history.last())
+    }
+
+    fun getAllRound(): List<Cars> = history.toList()
```
Comment (puleugo) [inline src/main/kotlin/racingcar/domain/race/RaceHistory.kt:15]: The method name doesn't clearly reflect its return type.
It might be helpful to rename it so that the name better matches the kind of value it returns.

File: src/main/kotlin/racingcar/util/StringParser.kt [hunk 5]
```diff
@@ -0,0 +1,18 @@
+package racingcar.util
+
+class StringParser private constructor() {
```
Comment (puleugo) [inline src/main/kotlin/racingcar/util/StringParser.kt:3]: I'm not fully convinced of the necessity of this class at this stage.
To be honest, I can agree that a`StringParser` class might be helpful later on, but with the current method implementation, it's a bit difficult to see the clear benefit right now.
It slightly increases the amount of code to read in order to understand the business logic.

File: src/main/kotlin/racingcar/controller/RacingController.kt [hunk 6]
```diff
@@ -0,0 +1,49 @@
...
+        val result = race.play()
+        outputView.showResult(race.getAllRounds())
+        return result
+    }
+
+    private fun determineWinner(result: RaceResult) {
+        val referee = Referee(result)
+        val winner = referee.determineWinner()
+        outputView.showWinner(winner)
+    }
```
Comment (jsoonworld) [inline src/main/kotlin/racingcar/controller/RacingController.kt:48]: I noticed that you're creating a `Referee` instance directly inside the `determineWinner()` method—was there a specific reason you chose to structure it this way? 😊

I was wondering if it might be worth considering letting `RaceResult` either determine the winner itself or internally create a `Referee` to handle that logic.

This could help reduce the controller's responsibility and bring the winner determination logic closer to where the data lives, potentially improving cohesion.
Just curious to hear your thoughts on this design decision! 🙌

File: src/main/kotlin/racingcar/controller/RacingController.kt [hunk 7]
```diff
@@ -0,0 +1,49 @@
...
+        val totalRounds = inputView.readTotalRounds()
+        return Round(totalRounds)
+    }
+
+    private fun playRace(round: Round, cars: Cars): RaceResult {
+        val race = Race(round, cars)
+        val result = race.play()
+        outputView.showResult(race.getAllRounds())
+        return result
+    }
```
Comment (jsoonworld) [inline src/main/kotlin/racingcar/controller/RacingController.kt:42]: I really like how you structured the `Race` object to encapsulate both race progression and result tracking!
That said, since `Race` is currently handling both game logic (`play`) and state/history management (`getAllRounds`),
I was wondering if it might be a bit heavy in terms of responsibility.

Just an idea—but would you consider separating this into dedicated components like `RaceProgressRecorder` or `RaceResultGenerator`?
It could help align the design more closely with the Single Responsibility Principle and make things easier to extend in the future.
Of course, your current design is also clean and clear—just sharing an alternative thought! 😊

File: src/main/kotlin/racingcar/domain/Referee.kt [hunk 8]
```diff
@@ -0,0 +1,11 @@
...
+
+import racingcar.domain.race.RaceResult
+
+class Referee(private val raceResult: RaceResult) {
+
+    fun determineWinner(): List<String> {
+        return raceResult.findCarsAtMaxPosition()
+            .map { it.getName() }
+    }
+}
```
Comment (jsoonworld) [inline src/main/kotlin/racingcar/domain/Referee.kt:11]: I really liked how you separated the responsibility of determining the winner into the `Referee` class!
Having the domain layer handle this logic instead of the controller makes the responsibilities feel well-separated and clear 😊

That said, the current implementation seems a bit minimal—it's essentially just calling `raceResult.findCarsAtMaxPosition()` and mapping to names.
So I was wondering: have you considered moving this logic into `RaceResult` itself as something like `raceResult.determineWinner()`?
Alternatively, if `Car` were to provide more data than just the name, expanding the role of Referee could also make sense.

Also, I was curious about the return type. Would returning `List<Name>` (as a Value Object) instead of `List<String>` be more expressive from a domain modeling perspective?

Just wanted to hear your thoughts on how you approached this—thank you! 🙏

File: src/main/kotlin/racingcar/domain/car/Car.kt [hunk 9]
```diff
@@ -0,0 +1,48 @@
...
+        return numberGenerator.generate() >= MOVEMENT_THRESHOLD
+    }
+
+    fun getName(): String {
+        return name.value
+    }
+
+    fun getPosition(): Int {
+        return position.value
+    }
```
Comment (jsoonworld) [inline src/main/kotlin/racingcar/domain/car/Car.kt:30]: I noticed you’re using `getName()` and `getPosition()` here. While that’s totally fine and familiar from Java,
in Kotlin it's often more idiomatic to expose such values as properties using something like `val name: String`.

I had the same habit when I first moved from Java to Kotlin 😅 but found that using property-based access really improved readability.
Just a small suggestion that might make the code feel a bit more Kotlin-like!

File: src/main/kotlin/racingcar/domain/car/Car.kt [hunk 10]
```diff
@@ -0,0 +1,48 @@
...
+
+    override fun equals(other: Any?): Boolean {
+        if (this === other) return true
+        if (other !is Car) return false
+
+        if (name != other.name) return false
+
+        return true
+    }
```
Comment (jsoonworld) [inline src/main/kotlin/racingcar/domain/car/Car.kt:39]: The current `equals()` implementation compares only the `name`, which works well if that's the only identity-defining property in your domain.

That said, I was curious if you’ve considered using a `data class` or even a `value class` (if applicable) for this.
Of course, that depends on whether including `position` in equality would go against your domain rules!
Just wanted to hear your thoughts on this design choice 😊

File: src/main/kotlin/racingcar/domain/car/Car.kt [hunk 11]
```diff
@@ -0,0 +1,48 @@
...
+
+import racingcar.domain.numbergenerator.NumberGenerator
+import racingcar.domain.numbergenerator.RandomNumberGenerator
+
+class Car(
+    private val name: Name,
+    private val position: Position,
+    private val numberGenerator: NumberGenerator
+) {
+    constructor(name: Name) : this(name, Position(), RandomNumberGenerator())
```
Comment (jsoonworld) [inline src/main/kotlin/racingcar/domain/car/Car.kt:11]: I really liked how you provided a secondary constructor for convenience!
One small thought—since you're instantiating `RandomNumberGenerator()` directly here,
it might be slightly harder to swap in a different implementation during testing or configuration.

Would it make sense to delegate this responsibility to a factory or inject it from outside?
That could give you a bit more flexibility down the road, especially if the number generation logic needs to change.
Just something to consider as the project evolves!

File: src/main/kotlin/racingcar/domain/car/Position.kt [hunk 12]
```diff
@@ -0,0 +1,14 @@
+package racingcar.domain.car
+
+class Position(val value: Int) {
```
Comment (jsoonworld) [inline src/main/kotlin/racingcar/domain/car/Position.kt:3]: One small thing I noticed—currently it's possible to create a `Position` with a negative value (e.g., `Position(-3)`).
If your domain assumes that positions should always be zero or positive, it might be helpful to add a simple guard clause in the constructor:
```kotlin
init {
    require(value >= 0) { "[ERROR] Position cannot be negative." }
}
```

This would prevent invalid states from being introduced and make the domain constraint more explicit at the object level.
Just something to consider if you want to enforce the rule more strictly! 😊

File: src/main/kotlin/racingcar/domain/numbergenerator/NumberGenerator.kt [hunk 13]
```diff
@@ -0,0 +1,6 @@
+package racingcar.domain.numbergenerator
+
+fun interface NumberGenerator {
```
Comment (jsoonworld) [inline src/main/kotlin/racingcar/domain/numbergenerator/NumberGenerator.kt:3]: I really loved how you abstracted the number generation logic into the `NumberGenerator` interface!
The use of a `fun interface` makes it so Kotlin-idiomatic and test-friendly 👏

File: src/main/kotlin/racingcar/domain/numbergenerator/RandomNumberGenerator.kt [hunk 14]
```diff
@@ -0,0 +1,15 @@
+package racingcar.domain.numbergenerator
+
+import camp.nextstep.edu.missionutils.Randoms
+
+class RandomNumberGenerator : NumberGenerator {
```
Comment (jsoonworld) [inline src/main/kotlin/racingcar/domain/numbergenerator/RandomNumberGenerator.kt:5]: This implementation is super clean—it's great to see how the actual logic is encapsulated away from the domain and neatly injected.
Definitely makes the whole design feel more modular and maintainable 🙌

File: src/main/kotlin/racingcar/domain/race/Race.kt [hunk 15]
```diff
@@ -0,0 +1,20 @@
+package racingcar.domain.race
+
+class Race(private val round: Round, private val cars: Cars) {
+
+    private val raceHistory = RaceHistory()
```
Comment (jsoonworld) [inline src/main/kotlin/racingcar/domain/race/Race.kt:5]: I noticed that `RaceHistory` is being directly instantiated within the Race class.
While this is totally fine for now, it might make things a bit harder to test or extend in the future—especially if `RaceHistory` needs to evolve or start depending on other components.

Have you considered injecting it via the constructor or introducing a factory method?
That could make it easier to swap out or mock in tests, and might give you more flexibility as the domain logic grows.

Just sharing a small thought in case you’re looking to make this part more extensible later on—great job overall! 😊

File: src/test/kotlin/racingcar/domain/RefereeTest.kt [hunk 16]
```diff
@@ -0,0 +1,56 @@
...
+    fun `determine the winner when there is only one winner`() {
+        // given
+        val car1 = createCar("a", mutableListOf(1))
+        val car2 = createCar("b", mutableListOf(8))
+
+        val referee = createRefereeWithCars(car1, car2)
+        // when
+        val winner = referee.determineWinner()
+        // then
+        assertThat(winner.first()).isEqualTo("b")
```
Comment (jsoonworld) [inline src/test/kotlin/racingcar/domain/RefereeTest.kt:25]: I really like how the test names clearly describe the expected behavior!
It makes the tests easy to follow and understand at a glance—nicely done 😊

File: src/test/kotlin/racingcar/domain/numberGenerator/FixedNumberGenerator.kt [hunk 17]
```diff
@@ -0,0 +1,22 @@
+package racingcar.domain.numberGenerator
+
+import racingcar.domain.numbergenerator.NumberGenerator
+
+class FixedNumberGenerator(val numbers: MutableList<Int>) : NumberGenerator {
```
Comment (jsoonworld) [inline src/test/kotlin/racingcar/domain/numberGenerator/FixedNumberGenerator.kt:5]: Great use of `FixedNumberGenerator` in your tests—it makes the outcomes predictable and helps ensure test reliability.
Really thoughtful approach 👏

**Review by jsoonworld — COMMENTED** [review]
Amazing work on this mission—huge props to you for putting so much thought and care into the implementation! 🙌
Even though Kotlin might still feel new, your code reflects a deep understanding of both Kotlin idioms and object-oriented design.

I was especially impressed by:

- The clean abstraction of `NumberGenerator` as a strategy,
- The use of a first-class collection in `Cars`,
- The immutable design of domain objects like `Car`, `Name`, and `Position`,
- And the predictable testing setup using `FixedNumberGenerator`.

Your test coverage is excellent, and I really appreciated how thoroughly you handled both normal and edge cases.
While reading through your code, it was clear that everything was written with intention and care 💯

Most of all, I loved how each part of the code clearly communicated its purpose.
It made reviewing your work not only enjoyable but also inspiring—thank you for that 😊

Looking forward to seeing how you continue to grow from here. Truly fantastic work—well done! 👏🙏

File: src/main/kotlin/racingcar/domain/numbergenerator/RandomNumberGenerator.kt (hunk 14 above)
Comment (puleugo) [inline src/main/kotlin/racingcar/domain/numbergenerator/RandomNumberGenerator.kt:5]: Why did you decided to extract this class? if you want to control random number in test. you might not be necessary this class. `missionutils.Randoms` can already be controlled in tests.

File: src/main/kotlin/racingcar/domain/race/Cars.kt [hunk 18]
```diff
@@ -0,0 +1,34 @@
+package racingcar.domain.race
+
+import racingcar.domain.car.Car
+
+class Cars(val cars: List<Car>) {
```
Comment (puleugo) [inline src/main/kotlin/racingcar/domain/race/Cars.kt:5]: Could you consider encapsulating the raw list?

File: src/main/kotlin/racingcar/domain/race/Cars.kt [hunk 19]
```diff
@@ -0,0 +1,34 @@
...
+        return cars.distinct().size == cars.size
+    }
+
+    fun moveAll(): Cars {
+        return Cars(cars.map { it.move() })
+    }
+
+    fun getMaxPosition() = cars.maxOf { it.getPosition() }
+
+    fun findAllAtSamePosition(position: Int): List<Car> {
```
Comment (puleugo) [inline src/main/kotlin/racingcar/domain/race/Cars.kt:22]: I think rename this method to `findAllByPosition` for readability.

File: src/main/kotlin/racingcar/domain/race/Cars.kt [hunk 20]
```diff
@@ -0,0 +1,34 @@
+package racingcar.domain.race
+
+import racingcar.domain.car.Car
+
+class Cars(val cars: List<Car>) {
+
+    init {
+        require(cars.size in MIN_SIZE..MAX_SIZE) { SIZE_OUT_OF_RANGE_ERROR }
```
Comment (puleugo) [inline src/main/kotlin/racingcar/domain/race/Cars.kt:8]: 👍 Great work

File: src/main/kotlin/racingcar/domain/race/Race.kt [hunk 21]
```diff
@@ -0,0 +1,20 @@
...
+        var currentCars = cars
+
+        repeat(round.value) {
+            currentCars = currentCars.moveAll()
+            raceHistory.recordRound(currentCars)
+        }
+        return raceHistory.getRaceResult()
+    }
+
+    fun getAllRounds(): List<Cars> {
```
Comment (puleugo) [inline src/main/kotlin/racingcar/domain/race/Race.kt:17]: The method name doesn't clearly reflect its return type.
It might be helpful to rename it so that the name better matches the kind of value it returns.

File: src/main/kotlin/racingcar/domain/car/Position.kt [hunk 22]
```diff
@@ -0,0 +1,14 @@
...
+
+    constructor() : this(0)
```
Comment (puleugo) [inline src/main/kotlin/racingcar/domain/car/Position.kt:5]: What about extracting this into a static field?

```suggestion
    constructor() : this(START_POSITION)
```

File: src/main/kotlin/racingcar/domain/race/Round.kt [hunk 23]
```diff
@@ -0,0 +1,16 @@
...
+
+class Round(val value: Int) {
+
+    init {
+        require(value in MIN_ROUND..MAX_ROUND) { ROUND_OUT_OF_RANGE_ERROR }
+    }
+
+    companion object {
+        private const val MIN_ROUND: Int = 1
+        private const val MAX_ROUND: Int = 10
```
Comment (puleugo) [inline src/main/kotlin/racingcar/domain/race/Round.kt:11]: I think rename this field for improve readability: 

```suggestion
        private const val MIN_ROUND_LIMIT: Int = 1
        private const val MAX_ROUND_LIMIT: Int = 10
```

File: src/main/kotlin/racingcar/view/OutputView.kt [hunk 24]
```diff
@@ -0,0 +1,29 @@
+package racingcar.view
+
+import racingcar.domain.race.Cars
+
+class OutputView {
+
+    fun showResult(allRoundCars: List<Cars>) {
```
Comment (puleugo) [inline src/main/kotlin/racingcar/view/OutputView.kt:7]: Could you consider renaming the methods in `OutputView` to start with `print*`?
This would help clarify that these methods are related to outputting to a console, improving readability and intent.

**Review by puleugo — APPROVED** [review]
Sorry for the late review!

Overall, this looks great from an OOP perspective. Please just take a look at the following comments.

Also, for better readability and reviewability of code changes, would you consider using stacked PRs?
You can find a helpful explanation [here](https://en.puleugo.dev/creating-a-mutually-respectful-pr-7c8f3de829c6).

File: src/main/kotlin/racingcar/controller/RacingController.kt (hunk 7 above)
Comment (Jiihyun) [reply src/main/kotlin/racingcar/controller/RacingController.kt:42]: Yeah, that could be a good idea!
Thanks for suggesting a great alternative ☺️

The role of `RaceHistory` is to record the state of cars for each round and generate the results. By separating this responsibility into an external class, `Race` can focus solely on the core logic of running the race. 
Also, this separation can clarify the responsibilities of each class and improves testability!!

File: src/main/kotlin/racingcar/controller/RacingController.kt (hunk 6 above)
Comment (Jiihyun) [reply src/main/kotlin/racingcar/controller/RacingController.kt:48]: Thanks for your thoughtful question! 😊

I wanted the controller to focus on managing the program’s overall flow, so I delegated the responsibility of determining the winner to a `Referee` object. I felt that having `RaceResult` decide the winner didn’t align with its contextual role. It felt more appropriate to treat it as a DTO.

Also, similar to your `racePolicy` example, if the policy for determining the winner changes, I only need to modify the `Referee` object. So, I thought that this makes the design more flexible and easier to maintain over time!!

File: src/main/kotlin/racingcar/domain/Referee.kt (hunk 8 above)
Comment (Jiihyun) [reply src/main/kotlin/racingcar/domain/Referee.kt:11]: I totally agree with your suggestion. It’s a great idea!

My original intent was to use `RaceResult` as a DTO and have Referee handle the winner determination like this:
```kotlin
fun determineWinner(raceResult: RaceResult): List<String> {
    val cars = raceResult.raceCompletedCars
    val maxPosition = cars.getMaxPosition()
    return cars.findAllAtSamePosition(maxPosition)
        .map { it.getName() }
}
```

However, I ended up trying to avoid directly using getters in `RaceResult` and added methods to it instead. Because of that, as you pointed out, the `Referee` is now just calling `raceResult.findCarsAtMaxPosition()` and mapping to names, which feels redundant.

I think this part definitely needs refactoring.
Thank you for pointing it out! 🙌

File: src/main/kotlin/racingcar/domain/car/Car.kt (hunk 11 above)
Comment (Jiihyun) [reply src/main/kotlin/racingcar/domain/car/Car.kt:11]: Actually, I’ve set up `Car` to have a `NumberGenerator` interface as a field, which I think makes it more flexible. By depending on the interface rather than directly on `RandomNumberGenerator`, I can easily swap in a different implementation if needed, whether it’s for testing or updating the logic in the future.

For example, during testing, I could inject a `FixedNumberGenerator` to control the behavior without altering the actual Car logic. This approach aligns with keeping the code open for extension but closed for modification!!

File: src/main/kotlin/racingcar/domain/car/Car.kt (hunk 9 above)
Comment (Jiihyun) [reply src/main/kotlin/racingcar/domain/car/Car.kt:30]: Thanks for letting me know! 
I'll try to get into the habit of writing more Kotlin-style code 😊

File: src/main/kotlin/racingcar/domain/car/Car.kt (hunk 10 above)
Comment (Jiihyun) [reply src/main/kotlin/racingcar/domain/car/Car.kt:39]: I wanted to use a data class too, but since `Car` objects are considered the same if they only have the same name, I couldn't use it in this case 🥺

File: src/main/kotlin/racingcar/domain/car/Position.kt (hunk 12 above)
Comment (Jiihyun) [reply src/main/kotlin/racingcar/domain/car/Position.kt:3]: I've initialized `Position` to 0 using the secondary constructor, but I missed making the primary constructor private.
Thank you for pointing that out 😊

File: src/main/kotlin/racingcar/domain/numbergenerator/NumberGenerator.kt (hunk 13 above)
Comment (Jiihyun) [reply src/main/kotlin/racingcar/domain/numbergenerator/NumberGenerator.kt:3]: Ohhh, thank you so much for the kind words!

File: src/main/kotlin/racingcar/domain/race/Race.kt (hunk 15 above)
Comment (Jiihyun) [reply src/main/kotlin/racingcar/domain/race/Race.kt:5]: You're absolutely right. Injecting `RaceHistory` through the constructor or from an external source would definitely make the code more flexible.

Thanks for pointing that out! I'll work on refactoring it with that in mind.

File: src/test/kotlin/racingcar/domain/RefereeTest.kt (hunk 16 above)
Comment (Jiihyun) [reply src/test/kotlin/racingcar/domain/RefereeTest.kt:25]: Thanks for the compliment! Naming variables and test cases are surprisingly tricky for me, so I'm glad to hear you are easy to follow haha

**Review by Jiihyun — COMMENTED** [review]
Hi! Thank you for your detailed and thorough review. 🥺

Thanks to your feedback, I was able to learn a lot about areas for improvement and gain new perspectives. I really appreciate it!

Good luck with your next mission as well! 😊

File: src/main/kotlin/racingcar/controller/RacingController.kt (hunk 3 above)
Comment (Jiihyun) [reply src/main/kotlin/racingcar/controller/RacingController.kt:24]: Yeah, that naming is much more readable!!
Thanks for the recommending such a good naming!

File: src/main/kotlin/racingcar/controller/RacingController.kt (hunk 2 above)
Comment (Jiihyun) [reply src/main/kotlin/racingcar/controller/RacingController.kt:28]: Ohh you're right. I missed refactoring this!
Thanks for letting me know :)

File: src/main/kotlin/racingcar/domain/car/Position.kt (hunk 22 above)
Comment (Jiihyun) [reply src/main/kotlin/racingcar/domain/car/Position.kt:5]: That's a good idea👍🏻
I thought `0` also gives a clear meaning and doesn’t need to be made into a constant, but your recommendation is much more readable and clear!

File: src/main/kotlin/racingcar/domain/numbergenerator/RandomNumberGenerator.kt (hunk 14 above)
Comment (Jiihyun) [reply src/main/kotlin/racingcar/domain/numbergenerator/RandomNumberGenerator.kt:5]: To adhere to the Open-Close-Principle(OCP) and minimize code changes when the conditions for number generation modified!

If I were to use `missionutils.Randoms` directly, it could work for controlling random numbers in tests, but I wanted to go a step further to make the design more flexible and future-proof. By introducing the `NumberGenerator`interface and its `RandomNumberGenerator` and `FixedNumberGenerator` implementation, I ensured that the logic for generating numbers is fully abstracted and decoupled.

This approach allows me to swap in any number generation strategy—not just for testing but also for extending functionality in the future—without altering the core logic of the dependent classes. It aligns with the Open-Closed Principle and makes the codebase more modular and maintainable!!

File: src/main/kotlin/racingcar/domain/race/Cars.kt (hunk 18 above)
Comment (Jiihyun) [reply src/main/kotlin/racingcar/domain/race/Cars.kt:5]: I didn't fully understand your question. Could you explain it in more detail?
Currently, the `Cars` class is a first-class collection encapsulating a list!

File: src/main/kotlin/racingcar/domain/race/Cars.kt (hunk 19 above)
Comment (Jiihyun) [reply src/main/kotlin/racingcar/domain/race/Cars.kt:22]: Oh, that method name is much better. Thanks for recommending a good naming! 👍🏻

File: src/main/kotlin/racingcar/domain/race/RaceHistory.kt (hunk 4 above)
Comment (Jiihyun) [reply src/main/kotlin/racingcar/domain/race/RaceHistory.kt:15]: I agree..! Now that I think about it again, a method name like `getAllHistories` would be better.

File: src/main/kotlin/racingcar/util/StringParser.kt (hunk 1 above)
Comment (Jiihyun) [reply src/main/kotlin/racingcar/util/StringParser.kt:8]: For scalability, I received the delimiter as a parameter, but I didn’t know that default parameters could be used. 
Since the program currently only uses `,` as a delimiter, using a default parameter would make the code more concise and convenient. 
Thanks for letting me know!

File: src/main/kotlin/racingcar/util/StringParser.kt (hunk 5 above)
Comment (Jiihyun) [reply src/main/kotlin/racingcar/util/StringParser.kt:3]: I wanted to separate responsibilities!
Since I only wanted `InputView` to handle user input, I created a `StringParser`class to take on the responsibility of parsing the user input.

By separating the responsibilities like this, I can gain several benefits.
1. **Easier Testing**: It becomes much easier to test individual components. You can now focus on testing the StringParser class independently from InputView, ensuring that parsing logic works correctly without worrying about user input.
2. **Flexibility & Scalability**: This separation also opens up the possibility to swap out or extend the parsing logic in the future (e.g., adding support for different formats or more complex parsing), without touching the user interface code in `InputView`.

File: src/main/kotlin/racingcar/view/OutputView.kt (hunk 24 above)
Comment (Jiihyun) [reply src/main/kotlin/racingcar/view/OutputView.kt:7]: I chose the `show*` naming convention for abstraction purposes!
I felt that "print" is too focused on console-based programs...🤔

**Review by Jiihyun — COMMENTED** [review]
Sorry for the late reply, as well.
Also thank you so much for the detail review. I really appreciate the time and effort you took to go through my code so thoroughly. Your feedback has been so helpful, and it’s given me a lot to think about and improve👍🏻

File: src/main/kotlin/racingcar/domain/race/Race.kt (hunk 21 above)
Comment (Jiihyun) [reply src/main/kotlin/racingcar/domain/race/Race.kt:17]: You're right. I might have considered introducing a data transfer object, such as a `RoundHistory`, to better represent the returned value and make the method's purpose more explicit!!
